nupic/research/frameworks/vernon/mixins/extra_validations.py

```diff
@@ -0,0 +1,140 @@
+
+from nupic.research.frameworks.pytorch.model_utils import aggregate_eval_results
+
+__all__ = [
+    "ExtraValidations"
+]
+
+
+class ExtraValidations:
+    """
+    Perform validations mid-epoch. Requires StepBasedLogging core mixin.
+    """
+    def setup_experiment(self, config):
+        """
+        :param config: Dictionary containing the configuration parameters
+
+            - extra_validations_per_epoch: number of additional validations to
+                                           perform mid-epoch. Additional
+                                           validations are distributed evenly
+                                           across training batches.
+        """
+        super().setup_experiment(config)
+
+        assert hasattr(self, "current_timestep"), (
+            "Must use StepBasedLogging or similar extension"
+        )
+
+        # A list of [(timestep, result), ...] for the current epoch.
+        self.extra_val_results = []
+
+        extra_validations = config.get("extra_validations_per_epoch", 0)
+        batches_to_validate = np.linspace(
+            min(self.total_batches, self.batches_in_epoch),
+            0,
+            1 + extra_validations,
+            endpoint=False
+        )[::-1].round().astype("int").tolist()
+        self.additional_batches_to_validate = batches_to_validate[:-1]
+        if extra_validations > 0:
+            self.logger.info(
+                f"Extra validations per epoch: {extra_validations}, "
+                f"batch indices: {self.additional_batches_to_validate}")
+
+    def run_epoch(self):
+        ret = super().run_epoch()
+        ret["extra_val_results"] = self.extra_val_results
+        self.extra_val_results = []
+        return ret
+
+    def post_batch(self, batch_idx, **kwargs):
+        super().post_batch(batch_idx=batch_idx, **kwargs)
+        validate = (batch_idx in self.additional_batches_to_validate
+                    and self.current_epoch in self.epochs_to_validate)
+        if validate:
+            result = self.validate()
+            self.extra_val_results.append(
+                (self.current_timestep, result)
+            )
+            self.model.train()
+
+    @classmethod
+    def expand_result_to_time_series(cls, result, config):
+        result_by_timestep = super().expand_result_to_time_series(result, config)
+
+        k_mapping = {
+            "mean_loss": "validation_loss",
+            "mean_accuracy": "validation_accuracy",
+            "learning_rate": "learning_rate",
+            "complexity_loss": "complexity_loss",
+        }
+        for timestep, val_result in result["extra_val_results"]:
+            result_by_timestep[timestep].update({
+                k2: val_result[k1]
+                for k1, k2 in k_mapping.items()
+                if k1 in val_result
+            })
+
+        return result_by_timestep
+
+    @classmethod
+    def aggregate_results(cls, results):
+        ret = super().aggregate_results(results)
```

Comment:
You include `if "aggregate_results" in eo` to pose the parent functionality as optional. However, here you assume the parent has the functionality. 

The fix is easy: `if hasattr(super(), aggregate_results)` 
But it's weird to think about its implications. This, in a sense, is optional inheritance. I'm not sure if this a red flag.

Maybe we can throw a warning when `ExtraValidations.aggregate_results` is called yet `super().aggregate_results` doesn't exist.

Reply:
Yeah, these "optional" methods were gross. I ended up doing a big refactor to get rid of them (creating `vernon.distributed` to capture all the distributed logic in subclass mixins)